overlord/devicestate/devicestate_install_mode_test.go

```diff
@@ -74,47 +100,84 @@ func (s *deviceMgrInstallModeSuite) makeMockInstalledPcGadget(c *C) {
 	snaptest.MockSnapWithFiles(c, "name: pc\ntype: gadget", si, [][]string{
 		{"meta/gadget.yaml", gadgetYaml},
 	})
-	s.makeModelAssertionInState(c, "canonical", "pc-model", map[string]interface{}{
+
+	si = &snap.SideInfo{
+		RealName: "core20",
+		Revision: snap.R(2),
+		SnapID:   core20SnapID,
+	}
+	snapstate.Set(s.state, "core20", &snapstate.SnapState{
+		SnapType: "base",
+		Sequence: []*snap.SideInfo{si},
+		Current:  si.Revision,
+		Active:   true,
+	})
+	snaptest.MockSnapWithFiles(c, "name: core20\ntype: base", si, nil)
+
+	s.mockModel = s.makeModelAssertionInState(c, "my-brand", "my-model", map[string]interface{}{
+		"display-name": "my model",
 		"architecture": "amd64",
-		"kernel":       "pc-kernel",
-		"gadget":       "pc",
-		"base":         "core18",
+		"base":         "core20",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              pcKernelSnapID,
+				"type":            "kernel",
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name":            "pc",
+				"id":              pcSnapID,
+				"type":            "gadget",
+				"default-channel": "20",
+			}},
 	})
 	devicestatetest.SetDevice(s.state, &auth.DeviceState{
-		Brand:  "canonical",
-		Model:  "pc-model",
-		Serial: "serial",
+		Brand: "my-brand",
+		Model: "my-model",
+		// no serial in install mode
 	})
 }
 
-func (s *deviceMgrInstallModeSuite) TestInstallModeCreatesChangeHappy(c *C) {
+func (s *deviceMgrInstallModeSuite) TestInstallModeRunChange(c *C) {
 	restore := release.MockOnClassic(false)
 	defer restore()
 
 	mockSnapBootstrapCmd := testutil.MockCommand(c, filepath.Join(dirs.DistroLibExecDir, "snap-bootstrap"), "")
 	defer mockSnapBootstrapCmd.Restore()
 
+	bootMakeBootableCalled := 0
+	restore = devicestate.MockBootMakeBootable(func(model *asserts.Model, rootdir string, bootWith *boot.BootableSet) error {
+		c.Check(model, DeepEquals, s.mockModel)
+		c.Check(rootdir, Equals, dirs.GlobalRootDir)
+		c.Check(bootWith.KernelPath, Matches, ".*/var/lib/snapd/snaps/pc-kernel_1.snap")
+		c.Check(bootWith.BasePath, Matches, ".*/var/lib/snapd/snaps/core20_2.snap")
+		bootMakeBootableCalled++
+		return nil
+	})
+	defer restore()
+
 	s.state.Lock()
+	defer s.state.Unlock()
```

Comment:
If you defer state unlocking here and an operation that runs in unlocked state (such as `s.settle()` below) fails it will panic with a double unlock.

Reply:
updated